### ISSUE TITLE
In-panel labs

### ISSUE THREAD

**Issue by baptiste** [rensa/stickylabeller#4]
Just the other day I needed a helper function to polish some plots for a paper: I often have facetted plots without facet strips to save on space, and the editors require a label `(a)`, `(b)`, etc. for each facet, which they see as a sub-figure.

It'd be nice to use your glue framework with this alternative labelling technique (a dummy text layer placed with the panels).

```
label_facets <- function(p, hjust=-0.5, vjust=1.5, fontface=2, ...){
  
  gb <- ggplot_build(p)
  lay <- gb$layout$panel_layout
  nm <- names(gb$layout$facet$params$rows)
  tags <- cbind(lay, label = paste0("(",letters[lay$PANEL],")"))
  p + geom_text(data=tags, aes(x = -Inf, y=Inf, label=label), ...,
            hjust=hjust, vjust=vjust, fontface=fontface, inherit.aes = FALSE) +
    theme(strip.text = element_blank(), strip.background = element_blank())
}
```

**Comment by rensa** [comment]
I think I understand the visual result you're describing, but maybe not the process you're thinking about using. Do you mean you'd:

* run the plot with facet strips in order to get their positions,
* grab the grobs for the facet strips,
* re-run the plot without strips so that the panels expand, and
* drop the facet strip grobs back in?

**Comment by baptiste** [comment]
I'll admit I didn't think this through very much... sorry for the confused description. I saw your package and thought it might be a good place to implement a solution for a related problem of tagging facets.


My common use-case is to remove strips and add tags to facets, 

<img width="675" alt="screen shot 2018-07-21 at 6 08 32 pm" src="https://user-images.githubusercontent.com/18970/43032680-24534312-8d11-11e8-8b34-e0d9df9c9af8.png">

Your package is targetting a somewhat different problem, but maybe there could be some overlap where the tag contains the same kind of information as the glued facets strips? Just a thought, feel free to close this issue.

```
mydf = data.frame(
  x = 1:90,
  y = rnorm(90),
  red = rep(letters[1:3], 30),
  blue = c(rep(1, 30), rep(2, 30), rep(3, 30)))

p <- ggplot(mydf) +
  geom_point(aes(x = x, y = y)) +
  facet_wrap(
    ~ red + blue)

label_facets <- function(p, hjust=-0.5, vjust=1.5, fontface=2, ...){
  
  gb <- ggplot_build(p)
  lay <- gb$layout$layout
  nm <- names(gb$layout$facet$params$rows)
  tags <- cbind(lay, label = paste0("(",letters[lay$PANEL],")"))
  p + geom_text(data=tags, aes(x = -Inf, y=Inf, label=label), ...,
                hjust=hjust, vjust=vjust, fontface=fontface, inherit.aes = FALSE) +
    theme(strip.text = element_blank(), strip.background = element_blank())
}

egg::ggarrange(p, label_facets(p), ncol=1)

```

**Comment by rensa** [comment]
No, that's okay! I was just looking around for alternate ways to solve the problem and saw that [at least one other person](https://stackoverflow.com/a/36090446/3246758) has taken a similar approach as you (use a dummy `geom_text` in lieu of facet labels).

I'd certainly prefer to find a solution that respects the semantic function of the facet labels, but if that's not possible then maybe this approach is the best way to make it happen 🙂

**Comment by rensa** [comment]
ggplot2 v3 also appears to have [a new `tag` argument in `labs()`](https://ggplot2.tidyverse.org/reference/labs.html), which seems designed to tackle a similar problem for the entire plot.

**Comment by rensa** [comment]
That seems like a shame given I see you can now choose whether strips go between the axis and the panel as well as choosing the side they sit on. I understand being opinionated about defaults, but if you're allowing customisability, I don't see a massive difference between that and allowing them inside the panel (except, perhaps, that the existing options still allocate space for the strips, rather than overlaying them).

**Comment by rensa** [comment]
Maybe a ggplot2 extension could achieve this as well?

**Comment by rensa** [comment]
So, I need to really sit down with this on the weekend a bit more, but I've been poking around the ggplot2 code and I'm wondering if creating modified facet classes is the way to do this. There's a lot going on, but as far as I can tell the `draw_panel` method in [`facet-wrap.r` (around line 320)](https://github.com/tidyverse/ggplot2/blob/510b4b4b3eae992255ddca86de63d1fada2221ca/R/facet-wrap.r#L320) and [`facet-grid-.r` (around line 368)](https://github.com/tidyverse/ggplot2/blob/510b4b4b3eae992255ddca86de63d1fada2221ca/R/facet-grid-.r#L368) start to consider the presence of strips w.r.t. facet panel dimensions. It's possible an alternate implementation could ignore this consideration, provided the labels were drawn on top of the panels and not underneath.

On the plus side, this would mean that facet labels could still be themed in the existing way. On the downside, this is a massively overengineered solution and changes in the facetting system in future versions of ggplot2 could mess it up.

**Comment by rensa** [comment]
Another downside: implementing it for `facet_grid` would probably be  a bad idea, since moving labels into the facets on the side would risk making it look like the labels only applied to them, instead of the entire row/column. I'm wondering if maybe we should just implement both, since this is a small package and it doesn't matter too much if the API has a couple of different approaches to the same problem 😄

**Comment by rensa** [comment]
Okay, I'm following the [Extending ggplot2](https://ggplot2.tidyverse.org/articles/extending-ggplot2.html#creating-a-new-faceting) vignette to try to sketch out how this would subclassing `facet_wrap`. My progress—which is really just the skeleton of a subclass and some notes about what I can try—are over on the [`feature-overlay` branch](https://github.com/rensa/stickylabeller/tree/feature-overlay)
.

**Comment by baptiste** [comment]
Cool, thanks for working on this.
Justification was never on the agenda for gtable, but kohske once wrote this draft for various workaround strategies (now 6 years old, things may have changed)

https://rpubs.com/kohske/815

**Comment by rensa** [comment]
Awesome 😁 I have the strip backgrounds going to the top (or bottom) successfully by moving the actual strip height from the parent `gTree` to the strip background rectGrob, using `unit(1, 'npc')` for the parent height and justifying the `rectGrob`. No need for additional dummy grobs.

The text is proving trickier, though, as the `textGrob` doesn't have a height, and it's wrapped in another gTree (a "titleGrob" `gTree`). I was hoping to use a similar strategy, making sure everything just expands to parent width, but no luck so far. I'll have a crack at the `vp` arguments 😄

**Comment by rensa** [comment]
Okay, this isn't _quite_ ready for prime-time, but it's getting there 😄 

```r
p = ggplot(mtcars) +
  geom_point(aes(x = mpg, y = gear)) +
  facet_wrap_overlay(~ carb, labeller = label_glue('({.l}) carb is {carb}')) +
  theme_grey(base_size = 10) +
  theme(
    strip.background = element_rect(fill = alpha('black', 0.5)),
    strip.text = element_text(colour = 'white', hjust = 0, vjust = 1.15)) +
  labs(
    title = 'Overlay facet titles in ggplot2 with stickylabeller!',
    subtitle = 'The facet_wrap_overlay function prints labels over the panels')
```

![overlay-test](https://user-images.githubusercontent.com/6520659/44000624-de508926-9e66-11e8-8b74-6ea989bd9183.png)

I got a bit frustrated with not being able to make things work with the text justification, so I ended up just doing the maths on its position. Maybe that can be fixed up in time 😁 But for now, it works with `strip.placement == c('top')` or `'left'`; `'bottom'` is missing the panels and `'right`' is off somewhere else altogether 😆 I'll do some more tweaking to get those going soon!
